python3.11-distorm3 - error during building

### Describe the bug
During the installation of Python3.11 Distorm3 package, I get:
```
============================= test session starts ==============================
platform linux -- Python 3.11.9, pytest-8.2.2, pluggy-1.5.0
rootdir: /build/source
go: downloading google.golang.org/protobuf v1.33.0
go: downloading github.com/go-git/gcfg v1.5.1-0.20230307220236-3a3c6141e376
collected 366 items / 3 errors                                                 

==================================== ERRORS ====================================
___________________ ERROR collecting python/test_distorm3.py ___________________
/nix/store/lqiim8h8978ysvnv4gwajxdmgyy3bxxl-python3.11-pytest-8.2.2/lib/python3.11/site-packages/_pytest/runner.py:341: in from_call
    result: Optional[TResult] = func()
/nix/store/lqiim8h8978ysvnv4gwajxdmgyy3bxxl-python3.11-pytest-8.2.2/lib/python3.11/site-packages/_pytest/runner.py:389: in collect
    return list(collector.collect())
/nix/store/lqiim8h8978ysvnv4gwajxdmgyy3bxxl-python3.11-pytest-8.2.2/lib/python3.11/site-packages/_pytest/unittest.py:90: in collect
    self.session._fixturemanager.parsefactories(self.newinstance(), self.nodeid)
/nix/store/lqiim8h8978ysvnv4gwajxdmgyy3bxxl-python3.11-pytest-8.2.2/lib/python3.11/site-packages/_pytest/unittest.py:75: in newinstance
    return self.obj("runTest")
E   TypeError: Test.__init__() takes 1 positional argument but 2 were given
___________________ ERROR collecting python/test_distorm3.py ___________________
/nix/store/lqiim8h8978ysvnv4gwajxdmgyy3bxxl-python3.11-pytest-8.2.2/lib/python3.11/site-packages/_pytest/runner.py:341: in from_call
    result: Optional[TResult] = func()
/nix/store/lqiim8h8978ysvnv4gwajxdmgyy3bxxl-python3.11-pytest-8.2.2/lib/python3.11/site-packages/_pytest/runner.py:389: in collect
    return list(collector.collect())
/nix/store/lqiim8h8978ysvnv4gwajxdmgyy3bxxl-python3.11-pytest-8.2.2/lib/python3.11/site-packages/_pytest/unittest.py:90: in collect
    self.session._fixturemanager.parsefactories(self.newinstance(), self.nodeid)
/nix/store/lqiim8h8978ysvnv4gwajxdmgyy3bxxl-python3.11-pytest-8.2.2/lib/python3.11/site-packages/_pytest/unittest.py:75: in newinstance
    return self.obj("runTest")
E   TypeError: InstBin.__init__() missing 3 required positional arguments: 'mode', 'features', and 'address'
___________________ ERROR collecting python/test_distorm3.py ___________________
/nix/store/lqiim8h8978ysvnv4gwajxdmgyy3bxxl-python3.11-pytest-8.2.2/lib/python3.11/site-packages/_pytest/runner.py:341: in from_call
    result: Optional[TResult] = func()
/nix/store/lqiim8h8978ysvnv4gwajxdmgyy3bxxl-python3.11-pytest-8.2.2/lib/python3.11/site-packages/_pytest/runner.py:389: in collect
    return list(collector.collect())
/nix/store/lqiim8h8978ysvnv4gwajxdmgyy3bxxl-python3.11-pytest-8.2.2/lib/python3.11/site-packages/_pytest/unittest.py:90: in collect
    self.session._fixturemanager.parsefactories(self.newinstance(), self.nodeid)
/nix/store/lqiim8h8978ysvnv4gwajxdmgyy3bxxl-python3.11-pytest-8.2.2/lib/python3.11/site-packages/_pytest/unittest.py:75: in newinstance
    return self.obj("runTest")
E   TypeError: Inst.__init__() missing 3 required positional arguments: 'mode', 'instNo', and 'features'
=========================== short test summary info ============================
ERROR python/test_distorm3.py::Test - TypeError: Test.__init__() takes 1 positional argument but 2 were given
ERROR python/test_distorm3.py::InstBin - TypeError: InstBin.__init__() missing 3 required positional arguments: 'mod...
ERROR python/test_distorm3.py::Inst - TypeError: Inst.__init__() missing 3 required positional arguments: 'mode',...
!!!!!!!!!!!!!!!!!!! Interrupted: 3 errors during collection !!!!!!!!!!!!!!!!!!!!
============================== 3 errors in 29.05s ==============================
/nix/store/5r0df66ikad3xw06azlqvswcvncll8wa-stdenv-linux/setup: line 1641: pop_var_context: head of shell_variables not a function context
go: downloading github.com/pjbgf/sha1cd v0.3.0
error: builder for '/nix/store/0mcd4xx6605j7fagjwqvafi03q9ni4sl-python3.11-distorm3-3.5.2.drv' failed with exit code 2
```

### Notify maintainers

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->
@fabaff 
### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.37, Athena OS, 24.11 (Vicuna), 24.11pre652179.7e7c39ea35c5`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.5`
 - channels(root): `"home-manager, nixos, nixpkgs"`
 - nixpkgs: `/home/athena/.nix-defexpr/channels/nixpkgs`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/python-modules/distorm3/default.nix
+++ b/pkgs/development/python-modules/distorm3/default.nix
@@ -1,8 +1,8 @@
 {
   lib,
   buildPythonPackage,
+  distutils,
   fetchFromGitHub,
-  pythonAtLeast,
   pythonOlder,
   pytestCheckHook,
   setuptools,
@@ -14,8 +14,7 @@ buildPythonPackage rec {
   version = "3.5.2";
   pyproject = true;
 
-  # Still uses distutils, https://github.com/gdabah/distorm/issues/191
-  disabled = pythonOlder "3.5" || pythonAtLeast "3.12";
+  disabled = pythonOlder "3.5";
 
   src = fetchFromGitHub {
     owner = "gdabah";
@@ -24,17 +23,18 @@ buildPythonPackage rec {
     hash = "sha256-Fhvxag2UN5wXEySP1n1pCahMQR/SfssywikeLmiASwQ=";
   };
 
-  build-system = [ setuptools ];
+  build-system = [
+    distutils
+    setuptools
+  ];
 
   nativeCheckInputs = [
     pytestCheckHook
     yasm
   ];
 
-  disabledTests = [
-    # TypeError: __init__() missing 3 required positional...
-    "test_dummy"
-  ];
+  # TypeError: __init__() missing 3 required positional...
+  doCheck = false;
 
   pythonImportsCheck = [ "distorm3" ];
 

--- a/pkgs/development/python-modules/distutils/default.nix
+++ b/pkgs/development/python-modules/distutils/default.nix
@@ -1,0 +1,54 @@
+{
+  lib,
+  buildPythonPackage,
+  fetchFromGitHub,
+  setuptools-scm,
+  setuptools,
+  python,
+  jaraco-envs,
+  jaraco-path,
+  jaraco-text,
+  more-itertools,
+  path,
+  pyfakefs,
+  pytestCheckHook,
+}:
+
+buildPythonPackage rec {
+  pname = "distutils";
+  inherit (setuptools) version;
+  pyproject = true;
+
+  src = fetchFromGitHub {
+    owner = "pypa";
+    repo = "distutils";
+    rev = "813ab9868b353991ec7324eb09619ee5beb9183f";
+    hash = "sha256-/YtITDuZlTJRisqsQ6SrgRRUrqLZpj+k3drrouURZlc=";
+  };
+
+  build-system = [ setuptools-scm ];
+
+  postInstall = ''
+    rm -r $out/${python.sitePackages}/distutils
+    ln -s ${setuptools}/${python.sitePackages}/setuptools/_distutils $out/${python.sitePackages}/distutils
+  '';
+
+  pythonImportsCheck = [ "distutils" ];
+
+  nativeCheckInputs = [
+    jaraco-envs
+    jaraco-path
+    jaraco-text
+    more-itertools
+    path
+    pyfakefs
+    pytestCheckHook
+  ];
+
+  meta = {
+    description = "Distutils as found in cpython";
+    homepage = "https://github.com/pypa/distutils";
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [ dotlambda ];
+  };
+}

--- a/pkgs/development/python-modules/jaraco-envs/default.nix
+++ b/pkgs/development/python-modules/jaraco-envs/default.nix
@@ -1,0 +1,48 @@
+{
+  lib,
+  buildPythonPackage,
+  fetchFromGitHub,
+  setuptools-scm,
+  path,
+  tox,
+  virtualenv,
+  pytestCheckHook,
+}:
+
+buildPythonPackage rec {
+  pname = "jaraco-envs";
+  version = "2.6.0";
+  pyproject = true;
+
+  src = fetchFromGitHub {
+    owner = "jaraco";
+    repo = "jaraco.envs";
+    rev = "refs/tags/v${version}";
+    hash = "sha256-yRMX0H6yWN8TiO/LGAr4HyrVS8ZhBjuR885/+UQscP0=";
+  };
+
+  build-system = [ setuptools-scm ];
+
+  dependencies = [
+    path
+    tox
+    virtualenv
+  ];
+
+  pythonImportsCheck = [ "jaraco.envs" ];
+
+  nativeCheckInputs = [ pytestCheckHook ];
+
+  disabledTestPaths = [
+    # requires networking
+    "jaraco/envs.py"
+  ];
+
+  meta = {
+    changelog = "https://github.com/jaraco/jaraco.envs/blob/${src.rev}/NEWS.rst";
+    description = "Classes for orchestrating Python (virtual) environments";
+    homepage = "https://github.com/jaraco/jaraco.envs";
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [ dotlambda ];
+  };
+}

--- a/pkgs/development/python-modules/jaraco-path/default.nix
+++ b/pkgs/development/python-modules/jaraco-path/default.nix
@@ -1,0 +1,34 @@
+{
+  lib,
+  buildPythonPackage,
+  fetchFromGitHub,
+  setuptools-scm,
+  pytestCheckHook,
+}:
+
+buildPythonPackage rec {
+  pname = "jaraco-path";
+  version = "3.7.0";
+  pyproject = true;
+
+  src = fetchFromGitHub {
+    owner = "jaraco";
+    repo = "jaraco.path";
+    rev = "refs/tags/v${version}";
+    hash = "sha256-P22sA138guKTlOxSxUJ0mU41W16984yYlkZea06juOM=";
+  };
+
+  build-system = [ setuptools-scm ];
+
+  pythonImportsCheck = [ "jaraco.path" ];
+
+  nativeCheckInputs = [ pytestCheckHook ];
+
+  meta = {
+    changelog = "https://github.com/jaraco/jaraco.path/blob/${src.rev}/NEWS.rst";
+    description = "Miscellaneous path functions";
+    homepage = "https://github.com/jaraco/jaraco.path";
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [ dotlambda ];
+  };
+}

--- a/pkgs/development/python-modules/python-ldap/default.nix
+++ b/pkgs/development/python-modules/python-ldap/default.nix
@@ -3,10 +3,10 @@
   stdenv,
   buildPythonPackage,
   fetchFromGitHub,
-  pythonAtLeast,
   pythonOlder,
 
   # build-system
+  distutils,
   setuptools,
 
   # native dependencies
@@ -26,7 +26,7 @@ buildPythonPackage rec {
   version = "3.4.4";
   pyproject = true;
 
-  disabled = pythonOlder "3.6" || pythonAtLeast "3.12"; # requires distutils
+  disabled = pythonOlder "3.6";
 
   src = fetchFromGitHub {
     owner = "python-ldap";
@@ -35,7 +35,10 @@ buildPythonPackage rec {
     hash = "sha256-v1cWoRGxbvvFnHqnwoIfmiQQcxfaA8Bf3+M5bE5PtuU=";
   };
 
-  build-system = [ setuptools ];
+  build-system = [
+    distutils
+    setuptools
+  ];
 
   buildInputs = [
     openldap

--- a/pkgs/development/python-modules/setuptools/default.nix
+++ b/pkgs/development/python-modules/setuptools/default.nix
@@ -2,6 +2,7 @@
   stdenv,
   lib,
   buildPythonPackage,
+  distutils,
   fetchFromGitHub,
   python,
   wheel,
@@ -32,6 +33,10 @@ buildPythonPackage rec {
 
   # Requires pytest, causing infinite recursion.
   doCheck = false;
+
+  passthru.tests = {
+    inherit distutils;
+  };
 
   meta = with lib; {
     description = "Utilities to facilitate the installation of Python packages";

--- a/pkgs/development/python-modules/thrift/default.nix
+++ b/pkgs/development/python-modules/thrift/default.nix
@@ -1,8 +1,8 @@
 {
   lib,
   buildPythonPackage,
+  distutils,
   fetchPypi,
-  pythonAtLeast,
   pythonOlder,
   setuptools,
   six,
@@ -13,15 +13,17 @@ buildPythonPackage rec {
   version = "0.20.0";
   pyproject = true;
 
-  # Still uses distutils
-  disabled = pythonOlder "3.7" || pythonAtLeast "3.12";
+  disabled = pythonOlder "3.7";
 
   src = fetchPypi {
     inherit pname version;
     hash = "sha256-TdZi6t9riuvopBcpUnvWmt9s6qKoaBy+9k0Sc7Po/ro=";
   };
 
-  build-system = [ setuptools ];
+  build-system = [
+    distutils
+    setuptools
+  ];
 
   dependencies = [ six ];
 

--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -6144,6 +6144,8 @@ self: super: with self; {
 
   jaraco-net = callPackage ../development/python-modules/jaraco-net { };
 
+  jaraco-path = callPackage ../development/python-modules/jaraco-path { };
+
   jaraco-stream = callPackage ../development/python-modules/jaraco-stream { };
 
   jaraco-test = callPackage ../development/python-modules/jaraco-test { };

--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -6132,6 +6132,8 @@ self: super: with self; {
 
   jaraco-email = callPackage ../development/python-modules/jaraco-email { };
 
+  jaraco-envs = callPackage ../development/python-modules/jaraco-envs { };
+
   jaraco-context = callPackage ../development/python-modules/jaraco-context { };
 
   jaraco-functools = callPackage ../development/python-modules/jaraco-functools { };

--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -3179,6 +3179,8 @@ self: super: with self; {
 
   distro = callPackage ../development/python-modules/distro { };
 
+  distutils = if pythonOlder "3.12" then null else callPackage ../development/python-modules/distutils { };
+
   distutils-extra = callPackage ../development/python-modules/distutils-extra { };
 
   # LTS in extended support phase

